File: 063/unique-paths-ii.go
func uniquePathsWithObstacles(obstacleGrid [][]int) int {
	m, n := len(obstacleGrid), len(obstacleGrid[0])

	if obstacleGrid[0][0] == 1 {
		return 0
	}
	if m == 1 {
		for _, i := range obstacleGrid[0] {
			if i == 1 {
				return 0
			}
		}
		return 1
	}

	matrix := make([][]int, m)
	for i := range matrix {
		row := make([]int, n)
		for j := range row {
			row[j] = 0
		}
		matrix[i] = row
	}

	for i := 0; i < m; i++ {
		if obstacleGrid[i][0] == 1 {
			for j := i; j < m; j++ {
				matrix[i][0] = 0
			}
			break
		} else {
			matrix[i][0] = 1
		}
	}

	for i := 0; i < n; i++ {
		if obstacleGrid[0][i] == 1 {
			for j := i; j < n; j++ {
				matrix[0][i] = 0
			}
			break
		} else {
			matrix[0][i] = 1
		}
	}

	for i := 1; i < m; i++ {
		for j := 1; j < n; j++ {
			if obstacleGrid[i][j] == 1 {
				matrix[i][j] = 0
			} else {
				matrix[i][j] = matrix[i-1][j] + matrix[i][j-1]
			}
		}
	}

	return matrix[m-1][n-1]
}

File: 063/UniquePathsII.kt
class UniquePathsII {

    fun uniquePathsWithObstacles(obstacleGrid: Array<IntArray>): Int {
        if (obstacleGrid[0][0] == 1) {
            return 0
        }
        if (obstacleGrid.size == 1) {
            return if (obstacleGrid[0].contains(1)) 0 else 1
        }

        val matrix = Array(obstacleGrid.size) { IntArray(obstacleGrid[0].size, { 0 }) }

        obstacleGrid.indices
                .takeWhile {
                    obstacleGrid[it][0] != 1
                }
                .forEach {
                    matrix[it][0] = 1
                }

        (0 until obstacleGrid[0].size)
                .takeWhile {
                    obstacleGrid[0][it] != 1
                }
                .forEach {
                    matrix[0][it] = if (obstacleGrid[0][it] == 1) 0 else 1
                }

        for (i in 1 until matrix.size) {
            for (j in 1 until matrix[0].size) {
                matrix[i][j] = if (obstacleGrid[i][j] == 1) 0 else matrix[i - 1][j] + matrix[i][j - 1]
            }
        }

        return matrix[matrix.size - 1][matrix[0].size - 1]
    }

}